From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: osy <osy@turing.llc>
Date: Fri, 4 Mar 2022 16:44:20 -0800
Subject: [PATCH 02/21] spice-gtk: user specified GMainContext for events

Following the previous commit, this replaces all GLib calls that
implicitly uses the default main context with versions that can use the
main context set by spice_util_set_main_context().

[spice-mac: ported from UTM patches/spice-gtk-0.42.patch to 0.43]
---
 src/channel-display-gst.c   | 10 +++++-----
 src/channel-display-mjpeg.c |  6 +++---
 src/channel-display.c       |  6 +++---
 src/channel-main.c          | 22 +++++++++++-----------
 src/channel-usbredir.c      |  2 +-
 src/channel-webdav.c        |  6 +++---
 src/gio-coroutine.c         | 14 ++++++++------
 src/smartcard-manager.c     |  5 +++--
 src/spice-channel.c         | 14 +++++++-------
 src/spice-gstaudio.c        | 10 +++++-----
 src/spice-gtk-session.c     |  8 ++++----
 src/spice-session.c         |  8 ++++----
 src/spice-widget.c          |  9 +++++----
 src/usb-acl-helper.c        |  3 ++-
 src/usb-device-manager.c    |  3 ++-
 src/vmcstream.c             |  2 +-
 16 files changed, 67 insertions(+), 61 deletions(-)

diff --git a/src/channel-display-gst.c b/src/channel-display-gst.c
index e063853..f720ca6 100644
--- a/src/channel-display-gst.c
+++ b/src/channel-display-gst.c
@@ -301,13 +301,13 @@ static void schedule_frame(SpiceGstDecoder *decoder)
         }
 
         if (spice_mmtime_diff(gstframe->encoded_frame->mm_time, now) >= 0) {
-            decoder->timer_id = g_timeout_add(gstframe->encoded_frame->mm_time - now,
-                                              display_frame, decoder);
+            decoder->timer_id = g_spice_timeout_add(gstframe->encoded_frame->mm_time - now,
+                                                    display_frame, decoder);
         } else if (decoder->display_frame && !decoder->pending_samples) {
             /* Still attempt to display the least out of date frame so the
              * video is not completely frozen for an extended period of time.
              */
-            decoder->timer_id = g_timeout_add(0, display_frame, decoder);
+            decoder->timer_id = g_spice_timeout_add(0, display_frame, decoder);
         } else {
             SPICE_DEBUG("%s: rendering too late by %u ms (ts: %u, mmtime: %u), dropping",
                         __FUNCTION__, now - gstframe->encoded_frame->mm_time,
@@ -771,7 +771,7 @@ static void spice_gst_decoder_reschedule(VideoDecoder *video_decoder)
     g_mutex_unlock(&decoder->queues_mutex);
 
     if (timer_id != 0) {
-        g_source_remove(timer_id);
+        g_spice_source_remove(timer_id);
     }
     schedule_frame(decoder);
 }
@@ -791,7 +791,7 @@ static void spice_gst_decoder_destroy(VideoDecoder *video_decoder)
      * scheduled display_frame() call and drop the queued frames.
      */
     if (decoder->timer_id) {
-        g_source_remove(decoder->timer_id);
+        g_spice_source_remove(decoder->timer_id);
     }
     g_mutex_clear(&decoder->queues_mutex);
     g_queue_free_full(decoder->decoding_queue, (GDestroyNotify)free_gst_frame);
diff --git a/src/channel-display-mjpeg.c b/src/channel-display-mjpeg.c
index 79b6ee9..a2c909e 100644
--- a/src/channel-display-mjpeg.c
+++ b/src/channel-display-mjpeg.c
@@ -183,7 +183,7 @@ static void mjpeg_decoder_schedule(MJpegDecoder *decoder)
             if (spice_mmtime_diff(time, frame->mm_time) <= 0) {
                 guint32 d = frame->mm_time - time;
                 decoder->cur_frame = frame;
-                decoder->timer_id = g_timeout_add(d, mjpeg_decoder_decode_frame, decoder);
+                decoder->timer_id = g_spice_timeout_add(d, mjpeg_decoder_decode_frame, decoder);
                 break;
             }
 
@@ -206,7 +206,7 @@ static void spice_frame_unref_func(gpointer data, gpointer user_data)
 
 static void mjpeg_decoder_drop_queue(MJpegDecoder *decoder)
 {
-    g_clear_handle_id(&decoder->timer_id, g_source_remove);
+    g_clear_handle_id(&decoder->timer_id, g_spice_source_remove);
     g_clear_pointer(&decoder->cur_frame, spice_frame_free);
     g_queue_foreach(decoder->msgq, spice_frame_unref_func, NULL);
     g_queue_clear(decoder->msgq);
@@ -251,7 +251,7 @@ static void mjpeg_decoder_reschedule(VideoDecoder *video_decoder)
     MJpegDecoder *decoder = (MJpegDecoder*)video_decoder;
 
     SPICE_DEBUG("%s", __FUNCTION__);
-    g_clear_handle_id(&decoder->timer_id, g_source_remove);
+    g_clear_handle_id(&decoder->timer_id, g_spice_source_remove);
     mjpeg_decoder_schedule(decoder);
 }
 
diff --git a/src/channel-display.c b/src/channel-display.c
index 0f658c3..c895ff5 100644
--- a/src/channel-display.c
+++ b/src/channel-display.c
@@ -151,7 +151,7 @@ static void spice_display_channel_dispose(GObject *object)
 {
     SpiceDisplayChannelPrivate *c = SPICE_DISPLAY_CHANNEL(object)->priv;
 
-    g_clear_handle_id(&c->mark_false_event_id, g_source_remove);
+    g_clear_handle_id(&c->mark_false_event_id, g_spice_source_remove);
 
     for (int i = 0; i < c->scanout.num_planes; i++) {
         if (c->scanout.fd[i] >= 0) {
@@ -2006,7 +2006,7 @@ static void display_handle_surface_create(SpiceChannel *channel, SpiceMsgIn *in)
         SPICE_DEBUG("surface flags: %x", create->flags);
         surface->primary = true;
         create_canvas(channel, surface);
-        g_clear_handle_id(&c->mark_false_event_id, g_source_remove);
+        g_clear_handle_id(&c->mark_false_event_id, g_spice_source_remove);
     } else {
         surface->primary = false;
         create_canvas(channel, surface);
@@ -2045,7 +2045,7 @@ static void display_handle_surface_destroy(SpiceChannel *channel, SpiceMsgIn *in
         CHANNEL_DEBUG(channel, "%d: FIXME primary destroy, but is display really disabled?", id);
         /* this is done with a timeout in spicec as well, it's *ugly* */
         if (id != 0 && c->mark_false_event_id == 0) {
-            c->mark_false_event_id = g_timeout_add_seconds(1, display_mark_false, channel);
+            c->mark_false_event_id = g_spice_timeout_add_seconds(1, display_mark_false, channel);
         }
         c->primary = NULL;
         g_coroutine_signal_emit(channel, signals[SPICE_DISPLAY_PRIMARY_DESTROY], 0);
diff --git a/src/channel-main.c b/src/channel-main.c
index 5402a45..991e10f 100644
--- a/src/channel-main.c
+++ b/src/channel-main.c
@@ -359,11 +359,11 @@ static void spice_main_channel_dispose(GObject *obj)
 {
     SpiceMainChannelPrivate *c = SPICE_MAIN_CHANNEL(obj)->priv;
 
-    g_clear_handle_id(&c->timer_id, g_source_remove);
+    g_clear_handle_id(&c->timer_id, g_spice_source_remove);
 
-    g_clear_handle_id(&c->switch_host_delayed_id, g_source_remove);
+    g_clear_handle_id(&c->switch_host_delayed_id, g_spice_source_remove);
 
-    g_clear_handle_id(&c->migrate_delayed_id, g_source_remove);
+    g_clear_handle_id(&c->migrate_delayed_id, g_spice_source_remove);
 
     g_clear_pointer(&c->file_xfer_tasks, g_hash_table_unref);
     g_clear_pointer (&c->flushing, g_hash_table_unref);
@@ -1172,7 +1172,7 @@ gboolean spice_main_channel_send_monitor_config(SpiceMainChannel *channel)
     agent_msg_queue(channel, VD_AGENT_MONITORS_CONFIG, size, mon);
 
     spice_channel_wakeup(SPICE_CHANNEL(channel), FALSE);
-    g_clear_handle_id(&c->timer_id, g_source_remove);
+    g_clear_handle_id(&c->timer_id, g_spice_source_remove);
 
     return TRUE;
 }
@@ -1543,16 +1543,16 @@ static void update_display_timer(SpiceMainChannel *channel, guint seconds)
     SpiceMainChannelPrivate *c = channel->priv;
 
     if (c->timer_id)
-        g_source_remove(c->timer_id);
+        g_spice_source_remove(c->timer_id);
 
     if (seconds != 0) {
-        c->timer_id = g_timeout_add_seconds(seconds, timer_set_display, channel);
+        c->timer_id = g_spice_timeout_add_seconds(seconds, timer_set_display, channel);
     } else {
         /* We need to special case 0, as we want the callback to fire as soon
          * as possible. g_timeout_add_seconds(0) would set up a timer which would fire
          * at the next second boundary, which might be nearly 1 full second later.
          */
-        c->timer_id = g_timeout_add(0, timer_set_display, channel);
+        c->timer_id = g_spice_timeout_add(0, timer_set_display, channel);
     }
 
 }
@@ -1770,7 +1770,7 @@ static void main_handle_channels_list(SpiceChannel *channel, SpiceMsgIn *in)
         c->id = msg->channels[i].id;
         /* no need to explicitly switch to main context, since
            synchronous call is not needed. */
-        G_GNUC_UNUSED guint idle_id = g_idle_add((GSourceFunc)_channel_new, c);
+        G_GNUC_UNUSED guint idle_id = g_spice_idle_add((GSourceFunc)_channel_new, c);
     }
 }
 
@@ -2549,7 +2549,7 @@ static void main_handle_migrate_end(SpiceChannel *channel, SpiceMsgIn *in)
     g_return_if_fail(c->migrate_delayed_id == 0);
     g_return_if_fail(spice_channel_test_capability(channel, SPICE_MAIN_CAP_SEMI_SEAMLESS_MIGRATE));
 
-    c->migrate_delayed_id = g_idle_add(migrate_delayed, channel);
+    c->migrate_delayed_id = g_spice_idle_add(migrate_delayed, channel);
 }
 
 /* main context */
@@ -2593,7 +2593,7 @@ static void main_handle_migrate_switch_host(SpiceChannel *channel, SpiceMsgIn *i
 
     if (c->switch_host_delayed_id != 0) {
         g_warning("Switching host already in progress, aborting it");
-        g_warn_if_fail(g_source_remove(c->switch_host_delayed_id));
+        g_warn_if_fail(g_spice_source_remove(c->switch_host_delayed_id));
         c->switch_host_delayed_id = 0;
     }
 
@@ -2606,7 +2606,7 @@ static void main_handle_migrate_switch_host(SpiceChannel *channel, SpiceMsgIn *i
     spice_session_set_port(session, mig->port, FALSE);
     spice_session_set_port(session, mig->sport, TRUE);
 
-    c->switch_host_delayed_id = g_idle_add(switch_host_delayed, channel);
+    c->switch_host_delayed_id = g_spice_idle_add(switch_host_delayed, channel);
 }
 
 /* coroutine context */
diff --git a/src/channel-usbredir.c b/src/channel-usbredir.c
index fe28762..9916ced 100644
--- a/src/channel-usbredir.c
+++ b/src/channel-usbredir.c
@@ -709,7 +709,7 @@ static void usbredir_handle_msg(SpiceChannel *c, SpiceMsgIn *in)
         err_data.device = spice_usb_backend_device_ref(device);
         err_data.error = err;
         spice_usbredir_channel_unlock(channel);
-        G_GNUC_UNUSED guint idle_id = g_idle_add(device_error, &err_data);
+        G_GNUC_UNUSED guint idle_id = g_spice_idle_add(device_error, &err_data);
         coroutine_yield(NULL);
 
         spice_usb_backend_device_unref(err_data.device);
diff --git a/src/channel-webdav.c b/src/channel-webdav.c
index f6146e6..450b361 100644
--- a/src/channel-webdav.c
+++ b/src/channel-webdav.c
@@ -97,7 +97,7 @@ static void output_queue_free(OutputQueue *queue)
     g_queue_free_full(queue->queue, g_free);
     g_clear_object(&queue->output);
     if (queue->idle_id)
-        g_source_remove(queue->idle_id);
+        g_spice_source_remove(queue->idle_id);
     g_free(queue);
 }
 
@@ -118,7 +118,7 @@ static void output_queue_flush_cb(GObject *source_object,
         g_warning("error: %s", error->message);
 
     if (!q->idle_id)
-        q->idle_id = g_idle_add(output_queue_idle, q);
+        q->idle_id = g_spice_idle_add(output_queue_idle, q);
 }
 
 static gboolean output_queue_idle(gpointer user_data)
@@ -170,7 +170,7 @@ static void output_queue_push(OutputQueue *q, const guint8 *buf, gsize size,
     g_queue_push_tail(q->queue, e);
 
     if (!q->idle_id && !q->flushing)
-        q->idle_id = g_idle_add(output_queue_idle, q);
+        q->idle_id = g_spice_idle_add(output_queue_idle, q);
 }
 #endif
 
diff --git a/src/gio-coroutine.c b/src/gio-coroutine.c
index 17251e9..af554f3 100644
--- a/src/gio-coroutine.c
+++ b/src/gio-coroutine.c
@@ -20,6 +20,8 @@
 #include "config.h"
 
 #include "gio-coroutine.h"
+#include "spice-version.h"
+#include "spice-util-priv.h"
 
 typedef struct _GConditionWaitSource
 {
@@ -55,13 +57,13 @@ GIOCondition g_coroutine_socket_wait(GCoroutine *self,
 
     g_autoptr(GSource) src = g_socket_create_source(sock, cond | G_IO_HUP | G_IO_ERR | G_IO_NVAL, NULL);
     g_source_set_callback(src, (GSourceFunc)g_io_wait_helper, self, NULL);
-    self->wait_id = g_source_attach(src, NULL);
+    self->wait_id = g_source_attach(src, spice_main_context());
     ret = coroutine_yield(NULL);
 
     if (ret != NULL)
         val = *ret;
     else
-        g_source_remove(self->wait_id);
+        g_spice_source_remove(self->wait_id);
 
     self->wait_id = 0;
     return val;
@@ -74,7 +76,7 @@ void g_coroutine_condition_cancel(GCoroutine *coroutine)
     if (coroutine->condition_id == 0)
         return;
 
-    g_clear_handle_id(&coroutine->condition_id, g_source_remove);
+    g_clear_handle_id(&coroutine->condition_id, g_spice_source_remove);
 }
 
 void g_coroutine_wakeup(GCoroutine *coroutine)
@@ -162,7 +164,7 @@ gboolean g_coroutine_condition_wait(GCoroutine *self, GConditionWaitFunc func, g
     vsrc->func = func;
     vsrc->data = data;
 
-    self->condition_id = g_source_attach(src, NULL);
+    self->condition_id = g_source_attach(src, spice_main_context());
     g_source_set_callback(src, g_condition_wait_helper, self, NULL);
     coroutine_yield(NULL);
 
@@ -215,7 +217,7 @@ g_coroutine_signal_emit(gpointer instance, guint signal_id,
         g_signal_emit_valist(instance, signal_id, detail, data.var_args);
     } else {
         g_object_ref(instance);
-        G_GNUC_UNUSED guint idle_id = g_idle_add(emit_main_context, &data);
+        G_GNUC_UNUSED guint idle_id = g_spice_idle_add(emit_main_context, &data);
         coroutine_yield(NULL);
         g_warn_if_fail(data.notified);
         g_object_unref(instance);
@@ -252,7 +254,7 @@ void g_coroutine_object_notify(GObject *object,
         data.propname = (gpointer)property_name;
         data.notified = FALSE;
 
-        G_GNUC_UNUSED guint idle_id = g_idle_add(notify_main_context, &data);
+        G_GNUC_UNUSED guint idle_id = g_spice_idle_add(notify_main_context, &data);
 
         /* This switches to the system coroutine context, lets
          * the idle function run to dispatch the signal, and
diff --git a/src/smartcard-manager.c b/src/smartcard-manager.c
index 7b6fe99..53005d0 100644
--- a/src/smartcard-manager.c
+++ b/src/smartcard-manager.c
@@ -27,6 +27,7 @@
 #include "smartcard-manager.h"
 #include "smartcard-manager-priv.h"
 #include "spice-marshal.h"
+#include "spice-util-priv.h"
 
 /**
  * SECTION:smartcard-manager
@@ -110,7 +111,7 @@ static void spice_smartcard_manager_finalize(GObject *gobject)
     SpiceSmartcardManager *manager = SPICE_SMARTCARD_MANAGER(gobject);
     SpiceSmartcardManagerPrivate *priv = manager->priv;
 
-    g_clear_handle_id(&priv->monitor_id, g_source_remove);
+    g_clear_handle_id(&priv->monitor_id, g_spice_source_remove);
 
 #ifdef USE_SMARTCARD
     g_clear_pointer(&priv->software_reader, vreader_free);
@@ -360,7 +361,7 @@ static guint smartcard_monitor_add(SmartcardSourceFunc callback,
 
     source = smartcard_monitor_source_new();
     g_source_set_callback(source, (GSourceFunc)callback, user_data, NULL);
-    id = g_source_attach(source, NULL);
+    id = g_source_attach(source, spice_main_context());
     g_source_unref(source);
 
     return id;
diff --git a/src/spice-channel.c b/src/spice-channel.c
index 3ce5d6e..6c5c4e8 100644
--- a/src/spice-channel.c
+++ b/src/spice-channel.c
@@ -743,9 +743,9 @@ void spice_msg_out_send(SpiceMsgOut *out)
     if (was_empty && !c->xmit_queue_wakeup_id) {
         c->xmit_queue_wakeup_id =
             /* Use g_timeout_add_full so that can specify the priority */
-            g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                               spice_channel_idle_wakeup,
-                               out->channel, NULL);
+            g_spice_timeout_add_full(G_PRIORITY_HIGH, 0,
+                                     spice_channel_idle_wakeup,
+                                     out->channel, NULL);
     }
 
 end:
@@ -2763,7 +2763,7 @@ cleanup:
         c->event = SPICE_CHANNEL_ERROR_CONNECT;
     }
 
-    G_GNUC_UNUSED guint idle_id = g_idle_add(spice_channel_delayed_unref, channel);
+    G_GNUC_UNUSED guint idle_id = g_spice_idle_add(spice_channel_delayed_unref, channel);
     /* Co-routine exits now - the SpiceChannel object may no longer exist,
        so don't do anything else now unless you like SEGVs */
     return NULL;
@@ -2822,7 +2822,7 @@ static gboolean channel_connect(SpiceChannel *channel, gboolean tls)
     g_object_ref(G_OBJECT(channel)); /* Unref'd when co-routine exits */
 
     /* we connect in idle, to let previous coroutine exit, if present */
-    c->connect_delayed_id = g_idle_add(connect_delayed, channel);
+    c->connect_delayed_id = g_spice_idle_add(connect_delayed, channel);
 
     return true;
 }
@@ -2887,7 +2887,7 @@ static void channel_reset(SpiceChannel *channel, gboolean migrating)
     SpiceChannelPrivate *c = channel->priv;
 
     CHANNEL_DEBUG(channel, "channel reset");
-    g_clear_handle_id(&c->connect_delayed_id, g_source_remove);
+    g_clear_handle_id(&c->connect_delayed_id, g_spice_source_remove);
 
 #ifdef HAVE_SASL
     if (c->sasl_conn) {
@@ -2916,7 +2916,7 @@ static void channel_reset(SpiceChannel *channel, gboolean migrating)
     gboolean was_empty = g_queue_is_empty(&c->xmit_queue);
     g_queue_foreach(&c->xmit_queue, (GFunc)spice_msg_out_unref, NULL);
     g_queue_clear(&c->xmit_queue);
-    g_clear_handle_id(&c->xmit_queue_wakeup_id, g_source_remove);
+    g_clear_handle_id(&c->xmit_queue_wakeup_id, g_spice_source_remove);
     g_mutex_unlock(&c->xmit_queue_lock);
     spice_channel_flushed(channel, was_empty);
 
diff --git a/src/spice-gstaudio.c b/src/spice-gstaudio.c
index a2363b9..1f94cf5 100644
--- a/src/spice-gstaudio.c
+++ b/src/spice-gstaudio.c
@@ -24,7 +24,7 @@
 #include "spice-gstaudio.h"
 #include "spice-common.h"
 #include "spice-session.h"
-#include "spice-util.h"
+#include "spice-util-priv.h"
 
 struct stream {
     GstElement              *pipe;
@@ -78,7 +78,7 @@ static void spice_gstaudio_dispose(GObject *obj)
     p = gstaudio->priv;
 
     stream_dispose(&p->playback);
-    g_clear_handle_id(&p->rbus_watch_id, g_source_remove);
+    g_clear_handle_id(&p->rbus_watch_id, g_spice_source_remove);
     stream_dispose(&p->record);
 
     if (p->pchannel)
@@ -193,7 +193,7 @@ static void record_start(SpiceRecordChannel *channel, gint format, gint channels
         (p->record.rate != frequency ||
          p->record.channels != channels)) {
         gst_element_set_state(p->record.pipe, GST_STATE_NULL);
-        g_clear_handle_id(&p->rbus_watch_id, g_source_remove);
+        g_clear_handle_id(&p->rbus_watch_id, g_spice_source_remove);
         g_clear_pointer(&p->record.pipe, gst_object_unref);
     }
 
@@ -244,7 +244,7 @@ static void playback_stop(SpiceGstaudio *gstaudio)
 
     if (p->playback.pipe)
         gst_element_set_state(p->playback.pipe, GST_STATE_READY);
-    g_clear_handle_id(&p->mmtime_id, g_source_remove);
+    g_clear_handle_id(&p->mmtime_id, g_spice_source_remove);
 }
 
 static gboolean update_mmtime_timeout_cb(gpointer data)
@@ -319,7 +319,7 @@ cleanup:
 
     if (!p->playback.fake && p->mmtime_id == 0) {
         update_mmtime_timeout_cb(gstaudio);
-        p->mmtime_id = g_timeout_add_seconds(1, update_mmtime_timeout_cb, gstaudio);
+        p->mmtime_id = g_spice_timeout_add_seconds(1, update_mmtime_timeout_cb, gstaudio);
     }
 }
 
diff --git a/src/spice-gtk-session.c b/src/spice-gtk-session.c
index 4e9ffec..faea2fa 100644
--- a/src/spice-gtk-session.c
+++ b/src/spice-gtk-session.c
@@ -287,7 +287,7 @@ static void clipboard_release_delay_remove(SpiceGtkSession *self, guint selectio
         clipboard_release(self, selection);
     }
 
-    g_source_remove(s->clipboard_release_delay[selection]);
+    g_spice_source_remove(s->clipboard_release_delay[selection]);
     s->clipboard_release_delay[selection] = 0;
 }
 
@@ -861,7 +861,7 @@ static void clipboard_get(GtkClipboard *clipboard,
 
     ri.selection_data = selection_data;
     ri.info = info;
-    ri.loop = g_main_loop_new(NULL, FALSE);
+    ri.loop = g_main_loop_new(spice_main_context(), FALSE);
     ri.selection = selection;
     ri.self = self;
 
@@ -1527,8 +1527,8 @@ static void clipboard_release_delay(SpiceMainChannel *main, guint selection,
     rel->self = self;
     rel->selection = selection;
     s->clipboard_release_delay[selection] =
-        g_timeout_add_full(G_PRIORITY_DEFAULT, CLIPBOARD_RELEASE_DELAY,
-                           clipboard_release_timeout, rel, g_free);
+        g_spice_timeout_add_full(G_PRIORITY_DEFAULT, CLIPBOARD_RELEASE_DELAY,
+                                 clipboard_release_timeout, rel, g_free);
 
 }
 
diff --git a/src/spice-session.c b/src/spice-session.c
index 9f58301..211918f 100644
--- a/src/spice-session.c
+++ b/src/spice-session.c
@@ -1847,7 +1847,7 @@ end:
     g_clear_object(&s->migration);
 
     s->migrate_wait_init = FALSE;
-    g_clear_handle_id(&s->after_main_init, g_source_remove);
+    g_clear_handle_id(&s->after_main_init, g_spice_source_remove);
 
     spice_session_set_migration_state(session, SPICE_SESSION_MIGRATION_NONE);
 }
@@ -1920,7 +1920,7 @@ gboolean spice_session_migrate_after_main_init(SpiceSession *self)
     g_return_val_if_fail(s->after_main_init == 0, FALSE);
 
     s->migrate_wait_init = FALSE;
-    s->after_main_init = g_idle_add(after_main_init, self);
+    s->after_main_init = g_spice_idle_add(after_main_init, self);
 
     return TRUE;
 }
@@ -2013,7 +2013,7 @@ void spice_session_disconnect(SpiceSession *session)
         return;
 
     g_object_ref(session);
-    s->disconnecting = g_idle_add((GSourceFunc)session_disconnect_idle, session);
+    s->disconnecting = g_spice_idle_add((GSourceFunc)session_disconnect_idle, session);
 }
 
 /**
@@ -2252,7 +2252,7 @@ GSocketConnection* spice_session_channel_open_host(SpiceSession *session, SpiceC
     g_socket_client_set_enable_proxy(open_host.client, s->proxy != NULL);
     g_socket_client_set_timeout(open_host.client, SOCKET_TIMEOUT);
 
-    G_GNUC_UNUSED guint idle_id = g_idle_add(open_host_idle_cb, &open_host);
+    G_GNUC_UNUSED guint idle_id = g_spice_idle_add(open_host_idle_cb, &open_host);
     /* switch to main loop and wait for connection */
     coroutine_yield(NULL);
 
diff --git a/src/spice-widget.c b/src/spice-widget.c
index 9788367..28db434 100644
--- a/src/spice-widget.c
+++ b/src/spice-widget.c
@@ -55,6 +55,7 @@
 #include "spice-gtk-session-priv.h"
 #include "vncdisplaykeymap.h"
 #include "spice-grabsequence-priv.h"
+#include "spice-util-priv.h"
 
 
 /**
@@ -484,7 +485,7 @@ static void spice_display_dispose(GObject *obj)
     g_clear_object(&d->session);
     d->gtk_session = NULL;
 
-    g_clear_handle_id(&d->key_delayed_id, g_source_remove);
+    g_clear_handle_id(&d->key_delayed_id, g_spice_source_remove);
 
     G_OBJECT_CLASS(spice_display_parent_class)->dispose(obj);
 }
@@ -1570,7 +1571,7 @@ static void key_press_and_release(SpiceDisplay *display)
     spice_inputs_channel_key_press_and_release(d->inputs, d->key_delayed_scancode);
     d->key_delayed_scancode = 0;
 
-    g_clear_handle_id(&d->key_delayed_id, g_source_remove);
+    g_clear_handle_id(&d->key_delayed_id, g_spice_source_remove);
 }
 
 static gboolean key_press_delayed(gpointer data)
@@ -1584,7 +1585,7 @@ static gboolean key_press_delayed(gpointer data)
     spice_inputs_channel_key_press(d->inputs, d->key_delayed_scancode);
     d->key_delayed_scancode = 0;
 
-    g_clear_handle_id(&d->key_delayed_id, g_source_remove);
+    g_clear_handle_id(&d->key_delayed_id, g_spice_source_remove);
 
     return G_SOURCE_REMOVE;
 }
@@ -1635,7 +1636,7 @@ static void send_key(SpiceDisplay *display, int scancode, SendKeyType type, gboo
             d->keypress_delay != 0 &&
             !(d->key_state[i] & m)) {
             g_warn_if_fail(d->key_delayed_id == 0);
-            d->key_delayed_id = g_timeout_add(d->keypress_delay, key_press_delayed, display);
+            d->key_delayed_id = g_spice_timeout_add(d->keypress_delay, key_press_delayed, display);
             d->key_delayed_scancode = scancode;
         } else
             spice_inputs_channel_key_press(d->inputs, scancode);
diff --git a/src/usb-acl-helper.c b/src/usb-acl-helper.c
index 9a9e0f5..bc7dc87 100644
--- a/src/usb-acl-helper.c
+++ b/src/usb-acl-helper.c
@@ -25,6 +25,7 @@
 #include <string.h>
 
 #include "usb-acl-helper.h"
+#include "spice-util-priv.h"
 
 struct _SpiceUsbAclHelperPrivate {
     GTask *task;
@@ -207,7 +208,7 @@ void spice_usb_acl_helper_open_acl_async(SpiceUsbAclHelper *self,
         g_task_return_error(task, err);
         goto done;
     }
-    g_child_watch_add(helper_pid, helper_child_watch_cb, NULL);
+    g_spice_child_watch_add(helper_pid, helper_child_watch_cb, NULL);
 
     priv->in_ch = g_io_channel_unix_new(in);
     g_io_channel_set_close_on_unref(priv->in_ch, TRUE);
diff --git a/src/usb-device-manager.c b/src/usb-device-manager.c
index fef2046..a2edaf7 100644
--- a/src/usb-device-manager.c
+++ b/src/usb-device-manager.c
@@ -38,6 +38,7 @@
 #include "spice-client.h"
 #include "spice-marshal.h"
 #include "usb-device-manager-priv.h"
+#include "spice-util-priv.h"
 
 #include <glib/gi18n-lib.h>
 
@@ -862,7 +863,7 @@ static void spice_usb_device_manager_hotplug_cb(void *user_data,
     args->manager = g_object_ref(manager);
     args->device = spice_usb_backend_device_ref(dev);
     args->added = added;
-    G_GNUC_UNUSED guint idle_id = g_idle_add(spice_usb_device_manager_hotplug_idle_cb, args);
+    G_GNUC_UNUSED guint idle_id = g_spice_idle_add(spice_usb_device_manager_hotplug_idle_cb, args);
 }
 
 static void spice_usb_device_manager_channel_connect_cb(GObject *gobject,
diff --git a/src/vmcstream.c b/src/vmcstream.c
index 18a1764..f866bc4 100644
--- a/src/vmcstream.c
+++ b/src/vmcstream.c
@@ -162,7 +162,7 @@ spice_vmc_input_stream_co_data(SpiceVmcInputStream *self,
         cb_data = g_new(complete_in_idle_cb_data , 1);
         cb_data->task = g_object_ref(self->task);
         cb_data->pos = self->pos;
-        G_GNUC_UNUSED guint idle_id = g_idle_add(complete_in_idle_cb, cb_data);
+        G_GNUC_UNUSED guint idle_id = g_spice_idle_add(complete_in_idle_cb, cb_data);
 
         g_clear_object(&self->task);
     }
-- 
2.56.0

